src/main/scala/tech/cryptonomic/conseil/tezos/TezosNodeMultiFetch.scala

```diff
@@ -0,0 +1,164 @@
+  /** the input type*/
+  type In
+  /** the output type*/
+  type Out
+  /** the json representation type used*/
+  type JsonType
+
+  /** an effectful function from a collection of inputs `T[In]`
+    * to the collection of json, tupled with the corresponding input `T[(In, JsonType)]`
+    */
+  def fetchJsonBatch: Kleisli[F, T[In], T[(In, JsonType)]]
+
+  /** an effectful function that decodes the json value to an output `Out` */
+  def decodeJson: Kleisli[F, JsonType, Out]
+
+  /** using the combination of the provided functions we fetch all data with this general outline:
+    * 1. get json values from the node, given a traversable input collection
+    * 2. keep the correlation with the input and convert every json to a corresponding output
+    * 3. return the traversable collection containing paired input and output values, all wrapped in the effect F
+    */
+  def fetch(implicit app: Monad[F], tr: Traverse[T]): Kleisli[F, T[In], T[(In, Out)]] =
+    fetchJsonBatch.andThen(json => decodeJson.second[In].traverse(json))
+
+}
+
+object TezosNodeMultiFetch {
+
+  /* An alias used to define function constraints on the internal dependent types (i.e. `I`, `O`, `J`)
+   * which would be otherwise transparent in any function signature
+   */
+  private type Aux[F[_], T[_], I, O, J] = TezosNodeMultiFetch[F, T] {
+      type In = I
+      type Out = O
+      type JsonType = J
+    }
+
+  /** Adds an extra json-decoding operation for every json intermediate result, returning both the
+    * original output `O` and the new decoded output `O2` as a tuple
+    */
+  def addDecoding[F[_]: Apply, T[_], I, O, O2, J](multiFetch: Aux[F,T,I,O,J], additionalDecode: Kleisli[F, J, O2]) = new TezosNodeMultiFetch[F, T] {
+    type In = I
+    type Out = (O, O2)
+    type JsonType = J
+
+    def fetchJsonBatch = multiFetch.fetchJsonBatch
+
+    def decodeJson = multiFetch.decodeJson.product(additionalDecode)
+  }
+
+  /** Allows to add an additional decoder to a "multi-fetch" instance, by providing an extension method `decodeAlso(additionalDecode)`,
+    * subject to `F` having a `cats.Apply` instance
+    */
+  object Syntax {
+    implicit class MultiFetchOps[F[_], T[_], I, O, J](m: Aux[F, T, I, O, J]) {
+      def decodeAlso[O2](additionalDecode: Kleisli[F, J, O2])(implicit app: Apply[F]): Aux[F, T, I, (O, O2), J] = addDecoding(m, additionalDecode)
+    }
+  }
+
+}
+
+/** Defines intances of `TezosNodeMultiFetch` for block-related data */
+object BlocksMultiFetchingInstances extends LazyLogging {
+  import scala.concurrent.{Future, ExecutionContext}
+  import cats.instances.future._
+  import io.circe.parser.decode
+
+  /** a fetcher of blocks */
+  def blocksMultiFetch(
+    network: String,
+    node: TezosRPCInterface,
+    hashRef: BlockHash,
+    concurrency: Int
+  ) = new TezosNodeMultiFetch[Future, List] {
+    import JsonDecoders.Circe.Blocks._
+
+    type JsonType = String
+    type In = Int //offset from head
+    type Out = BlockData
```

Comment:
Oh, okay, so In is any additional information needed to compute with the incoming tezos node data, JsonType is the type of the incoming data, and output is the resultant objects we desire from the computation?

Reply:
yep